### PR TITLE
C language support in tests

### PR DESCRIPTION
**Pull request by aciidb0mb3r**

### CODE REVIEW

File: Sources/Commands/UserToolchain.swift [hunk 1]
```diff
@@ -25,10 +25,15 @@ struct UserToolchain: Toolchain {
     let sysroot: String?
 
 #if os(OSX)
+    var targetArgs: [String] {
+        return ["-target", "x86_64-apple-macosx10.10"]
+    }
+
     var platformArgs: [String] {
-        return ["-target", "x86_64-apple-macosx10.10", "-sdk", sysroot!]
+        return targetArgs + ["-sdk", sysroot!]
```
Comment (ddunbar) [inline Sources/Commands/UserToolchain.swift]: Why is this necessary?

Clang accepts `-target`, but it isn't really the standard way to pass things to it (the standard way is `-arch x86_64 -mmacosx-version-min=10.10`.

File: Sources/Build/Command.link(SwiftModule).swift [hunk 2]
```diff
@@ -48,14 +48,15 @@ extension Command {
...
-        let inputs = objects
-
         switch product.type {
         case .Library(.Static):
             args.append(outpath)
         case .Test:
             args += ["-module-name", product.name]
           #if os(OSX)
+            // Link all the Clang Module's objects too into XCTest executable on OSX since XCTest is available in ObjC.
+            objects += product.modules.flatMap{ $0 as? ClangModule }.flatMap{ ClangModuleBuildMetadata(module: $0, prefix: prefix, otherArgs: []).objects }
```
Comment (ddunbar) [inline Sources/Build/Command.link(SwiftModule).swift]: I'm not sure if this is the right place for the conditional, it seems like we should instead avoid loading the C language target test modules on non-OSX for now.

File: Sources/Commands/UserToolchain.swift (hunk 1 above)
Comment (aciidb0mb3r) [inline Sources/Commands/UserToolchain.swift]: We later link swift and C language objects, swift is compiled with that target so linker gives a warning otherwise.

File: Sources/Build/Command.link(SwiftModule).swift (hunk 2 above)
Comment (aciidb0mb3r) [inline Sources/Build/Command.link(SwiftModule).swift]: We can error out in package loading for linux but I think this should still be in OSX conditional because this is the only case where we are linking Swift and C objects together, for other modules we interpolate via modulemaps.

File: Sources/Commands/UserToolchain.swift (hunk 1 above)
Comment (ddunbar) [inline Sources/Commands/UserToolchain.swift]: Makes sense. Let's just use the canonical form of passing the arguments to Clang.

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

File: Sources/Commands/UserToolchain.swift [hunk 3]
```diff
@@ -25,10 +25,15 @@ struct UserToolchain: Toolchain {
     let sysroot: String?
 
 #if os(OSX)
+    var clangTargetArgs: [String] {
+        return ["-arch", "x86_64", "-mmacosx-version-min=10.10"]
+    }
```
Comment (ddunbar) [inline Sources/Commands/UserToolchain.swift]: We should pass `["-isysroot", sdkroot!]` too, to match the code below, I think?

File: Sources/PackageModel/Module.swift [hunk 4]
```diff
@@ -110,20 +110,22 @@ public class CModule: Module {
...
     public let providers: [SystemPackageProvider]?
-    public init(name: String, path: String, pkgConfig: String? = nil, providers: [SystemPackageProvider]? = nil) throws {
+    public init(name: String, path: String, isTest: Bool = false, pkgConfig: String? = nil, providers: [SystemPackageProvider]? = nil) throws {
         self.path = path
         self.pkgConfig = pkgConfig
         self.providers = providers
-        try super.init(name: name)
+        // FIXME: This is wrong, System modules should never be a test module, perhaps ClangModule
+        // can be refactored into direct subclass of Module.
+        try super.init(name: name, isTest: isTest)
```
Comment (ddunbar) [inline Sources/PackageModel/Module.swift]: What is the thing that is wrong here? Does anyone pass isTest: true?

File: Sources/PackageLoading/PackageExtensions.swift [hunk 5]
```diff
@@ -154,19 +154,26 @@ extension Package {
...
         
         if !cSources.isEmpty {
             guard swiftSources.isEmpty else { throw Module.Error.mixedSources(path) }
-            return try ClangModule(name: name, sources: Sources(paths: cSources, root: path))
+
+          #if os(Linux)
+            guard !isTest else {
+                fatalError("\(name): C family test modules are not supported on linux.")
+            }
+          #endif
```
Comment (ddunbar) [inline Sources/PackageLoading/PackageExtensions.swift]: This is unfortunate, because otherwise cross-platform packages will just fail to build. I would prefer that -- for now -- we simply automatically ignore these targets on Linux, and work on a real platform compatibility specification API (in PackageDescription).

File: Sources/Build/Command.link(SwiftModule).swift (hunk 2 above)
Comment (ddunbar) [inline Sources/Build/Command.link(SwiftModule).swift]: I still think we should manage this more explicitly, where we explicitly compute the dependencies of the tests somewhere, and we just exclude the C language modules on Linux.

File: Sources/PackageModel/Module.swift (hunk 4 above)
Comment (aciidb0mb3r) [inline Sources/PackageModel/Module.swift]: Its not passed by anyone but potentially it can be which would be wrong.

File: Sources/PackageLoading/PackageExtensions.swift (hunk 5 above)
Comment (aciidb0mb3r) [inline Sources/PackageLoading/PackageExtensions.swift]: Ah yes, that is definitely better

File: Sources/Commands/UserToolchain.swift (hunk 3 above)
Comment (aciidb0mb3r) [inline Sources/Commands/UserToolchain.swift]: It doesn't complain but I guess that'll be consistent

**Comment by aciidb0mb3r** [comment]
@ddunbar updated

File: Sources/PackageLoading/PackageExtensions.swift [hunk 6]
```diff
@@ -227,6 +227,10 @@ extension Package {
             //TODO and then we should prefix all modules with their package probably
             //Suffix 'Tests' to test product so the module name of linux executable don't collide with
             //main package, if present.
+            // FIXME: Ignore C language test modules on linux for now.
+          #if os(Linux)
+            let testModules = testModules.filter { !($0 is ClangModule) }
```
Comment (ddunbar) [inline Sources/PackageLoading/PackageExtensions.swift]: Let's print some kind of warning here, as opposed to just silently ignoring it. That will be really annoying for anyone using such a package on Linux, but that can be our incentive to fix it in a more sensible fashion.

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by ddunbar** [comment]
LGTM... We still need Xcodeproj support is that right?

**Comment by aciidb0mb3r** [comment]
The files do come up in generated xcode proj but filetype is wrong (`sourcecode.c.c`) so it doesn't build. I remember filing a bug for that, will try to look into it.
